packages/hemi-tunnel-actions/src/initiateWithdraw.ts

```diff
@@ -0,0 +1,232 @@
+      const hash = await writeContract(l2WalletClient, {
+        abi: l2BridgeAbi,
+        account,
+        address: l2Bridge,
+        // See https://github.com/ethereum-optimism/ecosystem/blob/8da00d3b9044dcb58558df28bae278b613562725/packages/sdk/src/adapters/eth-bridge.ts#L178
+        args: [l2TokenAddress, amount, 0, '0x'],
+        chain: l2Chain,
+        functionName: 'withdraw',
+        value,
+      }).catch(function (error) {
+        emitter.emit('user-signing-withdraw-error', error)
+      })
+
+      if (!hash) {
+        return
+      }
+
+      emitter.emit('user-signed-withdraw', hash)
+
+      const withdrawalReceipt = await l2PublicClient
+        .waitForTransactionReceipt({
+          hash,
+        })
+        .catch(function (err) {
+          emitter.emit('withdraw-failed', err)
+        })
+      if (!withdrawalReceipt) {
+        return
+      }
+
+      emitter.emit(
+        withdrawalReceipt.status === 'success'
+          ? 'withdraw-transaction-succeeded'
+          : 'withdraw-transaction-reverted',
+        withdrawalReceipt,
+      )
+    } catch (error) {
+      emitter.emit('unexpected-error', error as Error)
+    } finally {
+      emitter.emit('withdraw-settled')
+    }
+  }
+
+export const encodeInitiateWithdraw = ({
+  amount = BigInt(0),
+  l2TokenAddress,
+}: {
+  amount: bigint | undefined
+  l2TokenAddress: Address
+}) =>
+  encodeFunctionData({
+    abi: l2BridgeAbi,
+    // See https://github.com/ethereum-optimism/ecosystem/blob/8da00d3b9044dcb58558df28bae278b613562725/packages/sdk/src/adapters/eth-bridge.ts#L178
+    args: [l2TokenAddress, amount, 0, '0x'],
+    functionName: 'withdraw',
+  })
+
+export const initiateWithdrawEth = ({
+  account,
+  amount,
+  l2PublicClient,
+  ...args
+}: Omit<InitiateWithdraw, 'checkBalance' | 'value'>) =>
+  toPromiseEvent<WithdrawEvents>(
+    runInitiateWithdraw({
+      account,
+      amount,
+      async checkBalance() {
+        const balance = await hasEthBalance({
+          account,
+          publicClient: l2PublicClient,
+        })
+        // using >= because we need to consider gas fees
+        if (amount >= balance) {
+          return 'insufficient balance'
+        }
+        return undefined
+      },
+      l2PublicClient,
+      // withdrawing ETH requires to send ETH as value, so we add it here
+      value: amount,
+      ...args,
+    }),
+  )
+
+export const initiateWithdrawErc20 = ({
+  account,
+  amount,
+  l2PublicClient,
+  l2TokenAddress,
+  ...args
+}: Omit<InitiateWithdraw, 'checkBalance' | 'value'>) =>
+  toPromiseEvent<WithdrawEvents>(
+    runInitiateWithdraw({
+      account,
+      amount,
+      async checkBalance() {
+        const balance = await hasErc20Balance({
+          account,
+          publicClient: l2PublicClient,
+          tokenAddress: l2TokenAddress,
+        })
+        if (amount > balance) {
+          return 'insufficient balance'
+        }
```

Comment:
If amount === balance, we can execute the withdrawal. Can we also check if the user has enough ETH to pay the (guesstimated) fees?

Reply:
Yes, I think I inverted the conditions between initiateWithdrawEth and initiateWithdrawErc20 🤦🏽

Reply:
> Can we also check if the user has enough ETH to pay the (guesstimated) fees?

I can check it has balances

Reply:
Added check for balance here [9c7b658](https://github.com/hemilabs/ui-monorepo/pull/1135/commits/9c7b658968de1b317167195d350be0a0898e38f6)